### PR TITLE
SH-112 PU - User create API to take additional 'managedBy' param

### PR DESCRIPTION
**Pull request by reshmi-nair**
User create API to take additional 'managedBy' param

- “managedBy” should be UUID of a valid user and this user should have “managedBy” value as null
- rootOrg of the managed user (i.e. when managedBy is non-null) should be the custodian org

### CODE REVIEW

File: actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java [hunk 1]
```diff
@@ -617,19 +636,20 @@ private void processUserRequestV3(Map<String, Object> userMap, String signupType
...
-            (userMap.get(JsonKey.EMAIL_VERIFIED) != null
-                ? userMap.get(JsonKey.EMAIL_VERIFIED)
-                : false));
-    userFlagsMap.put(
-        JsonKey.PHONE_VERIFIED,
-        (Boolean)
-            (userMap.get(JsonKey.PHONE_VERIFIED) != null
-                ? userMap.get(JsonKey.PHONE_VERIFIED)
-                : false));
+    if (StringUtils.isEmpty(managedBy)) {
```
Comment (Hari-stackroute) [inline actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java:639]: If it is managed user, we can set emailVerified and phoneVerified to default values as false right?

File: actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java [hunk 2]
```diff
@@ -603,8 +612,18 @@ private void throwParameterMismatchException(String... param) {
...
     UserUtil.setUserDefaultValueForV3(userMap);
     UserUtil.toLower(userMap);
-    UserUtil.checkPhoneUniqueness((String) userMap.get(JsonKey.PHONE));
-    UserUtil.checkEmailUniqueness((String) userMap.get(JsonKey.EMAIL));
+
+    String managedBy = (String) userMap.get(JsonKey.MANAGED_BY);
+    if (StringUtils.isEmpty(managedBy)) {
+      UserUtil.checkPhoneUniqueness((String) userMap.get(JsonKey.PHONE));
+      UserUtil.checkEmailUniqueness((String) userMap.get(JsonKey.EMAIL));
+    }else{
```
Comment (Hari-stackroute) [inline actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java]: format this to "} else {"

Comment (reshmi-nair) [reply actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java]: Done the changes

File: actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java [hunk 3]
```diff
@@ -831,6 +862,10 @@ public Response apply(String parameter) {
     }
     requestMap.put(JsonKey.PASSWORD, userMap.get(JsonKey.PASSWORD));
     if (StringUtils.isNotBlank(callerId)) {
+      if (StringUtils.isNotEmpty(managedBy)) {
+        requestMap.put(JsonKey.EMAIL, managedByEmail);
```
Comment (indrajra) [inline actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java:866]: this is used to send notification about account creation. But seems like the map is carrying this as if the managed user has it. Dislike this idea.

File: actors/user/src/main/java/org/sunbird/user/util/UserUtil.java [hunk 4]
```diff
@@ -760,6 +760,23 @@ public static void toLower(Map<String, Object> userMap) {
...
+    if (ProjectUtil.isNull(managedByInfo)
+            || StringUtils.isBlank((String) managedByInfo.get(JsonKey.FIRST_NAME))
+            || StringUtils.isNotBlank((String) managedByInfo.get(JsonKey.MANAGED_BY))
+            || (null != managedByInfo.get(JsonKey.IS_DELETED) && (boolean) (managedByInfo.get(JsonKey.IS_DELETED))) ) {
+      throw new ProjectCommonException(
+              ResponseCode.invalidRequestData.getErrorCode(),
+              ResponseCode.invalidRequestData.getErrorMessage(),
+              ResponseCode.CLIENT_ERROR.getResponseCode());
+    }
+    UserUtility.decryptUserDataFrmES(managedByInfo);
```
Comment (indrajra) [inline actors/user/src/main/java/org/sunbird/user/util/UserUtil.java:777]: decryption is done to get email or phone in plain text - to later send notifications.

**Pull request merged**